pkg/payload/task_graph.go

```diff
@@ -437,169 +435,117 @@ func RunGraph(ctx context.Context, graph *TaskGraph, maxParallelism int, fn func
-				if finished.success {
-					visited[finished.index] = nodeComplete
-				} else {
-					visited[finished.index] = nodeFailed
-				}
-				remaining--
-				inflight--
-				found++
-			}
-
-			if found > 0 {
+	getNextNode := func() int {
+		for i, node := range graph.Nodes {
+			if submitted[i] {
 				continue
 			}
-
-			// no more work to hand out
-			if remaining == 0 {
-				klog.V(4).Infof("Graph is complete")
-				return
+			if canVisit(node) {
+				return i
 			}
+		}
 
-			// we walked the entire graph, there are still nodes remaining, but we're not waiting
-			// for anything
-			if inflight == 0 && found == 0 {
-				klog.V(4).Infof("No more reachable nodes in graph, continue")
-				break
-			}
+		return -1
+	}
 
-			// we did nothing this round, so we have to wait for more
-			finished, ok := <-completeCh
-			if !ok {
-				// we've been aborted
-				klog.V(4).Infof("Stopped graph walker due to cancel")
-				return
-			}
-			if finished.success {
-				visited[finished.index] = nodeComplete
-			} else {
-				visited[finished.index] = nodeFailed
-			}
-			remaining--
-			inflight--
-		}
+	// Tasks go out to the workers via workCh, and results come brack
+	// from the workers via resultCh.
+	workCh := make(chan runTasks, maxParallelism)
+	defer close(workCh)
 
-		// take everything remaining and process in order
-		var unreachable []*Task
-		for i := 0; i < len(visited); i++ {
-			if visited[i] == nodeNotVisited && canVisit(graph.Nodes[i]) {
-				unreachable = append(unreachable, graph.Nodes[i].Tasks...)
-			}
-		}
-		if len(unreachable) > 0 {
-			sort.Slice(unreachable, func(i, j int) bool {
-				a, b := unreachable[i], unreachable[j]
-				return a.Index < b.Index
-			})
-			workCh <- runTasks{index: -1, tasks: unreachable}
-			klog.V(4).Infof("Waiting for last tasks")
-			<-completeCh
-		}
-		klog.V(4).Infof("No more work")
-	}()
+	resultCh := make(chan taskStatus, maxParallelism)
+	defer close(resultCh)
 
-	errCh := make(chan error, maxParallelism)
 	wg := sync.WaitGroup{}
 	if maxParallelism < 1 {
 		maxParallelism = 1
 	}
 	for i := 0; i < maxParallelism; i++ {
 		wg.Add(1)
-		go func(job int) {
+		go func(ctx context.Context, job int) {
 			defer utilruntime.HandleCrash()
 			defer wg.Done()
 			for {
 				select {
-				case <-nestedCtx.Done():
-					klog.V(4).Infof("Canceled worker %d", job)
+				case <-ctx.Done():
+					klog.V(4).Infof("Canceled worker %d while waiting for work", job)
 					return
-				case runTask, ok := <-workCh:
-					if !ok {
-						klog.V(4).Infof("No more work for %d", job)
-						return
-					}
+				case runTask := <-workCh:
 					klog.V(4).Infof("Running %d on %d", runTask.index, job)
-					err := fn(nestedCtx, runTask.tasks)
-					completeCh <- taskStatus{index: runTask.index, success: err == nil}
-					if err != nil {
-						errCh <- err
-					}
+					err := fn(ctx, runTask.tasks)
+					resultCh <- taskStatus{index: runTask.index, error: err}
 				}
 			}
-		}(i)
+		}(nestedCtx, i)
```

Comment:
do we need this nextedCtx ?

Reply:
hmm.. i see we use it to shutdown workers when the Sync is done or canceled.

could we move this nextexCtx definition closer to these worker start instead of top of Sync?

Reply:
> could we move this nextexCtx definition closer to these worker start instead of top of Sync?

We may have to shift it back in the future if `RunGraph` grows something else it needs to shut down in the future.  But sure, shifted closer to worker start in 7f95e2e -> f6271d0.